src/components/modals/FlipInputModal2.tsx

```diff
@@ -0,0 +1,301 @@
+  onMaxSet?: () => void
+  onAmountsChanged?: (amounts: ExchangedFlipInputAmounts) => void
+  headerText?: string
+  hideMaxButton?: boolean
+}
+
+const FlipInputModal2Component = React.forwardRef<FlipInputModalRef, Props>((props: Props, ref) => {
+  const {
+    bridge,
+    walletId,
+    tokenId,
+    startNativeAmount,
+    forceField,
+    onAmountsChanged,
+    feeTokenId: startingFeeTokenId,
+    feeNativeAmount: startingFeeNativeAmount = '',
+    feeStyle,
+    onFeesChange,
+    onMaxSet,
+    headerText,
+    hideMaxButton
+  } = props
+
+  const account = useSelector(state => state.core.account)
+  const currencyWallets = useWatch(account, 'currencyWallets')
+  const [feeTokenId, setFeeTokenId] = useState<string | undefined>(startingFeeTokenId)
+  const [feeNativeAmount, setFeeNativeAmount] = useState<string>(startingFeeNativeAmount)
+  const [errorMessage, setErrorMessage] = useState<string | null>(null)
+  const wallet = currencyWallets[walletId]
+  const { currencyInfo } = wallet
+  const balances = useWatch(wallet, 'balances')
+  const { pluginId } = currencyInfo
+  const currencyCode = getCurrencyCode(wallet, tokenId)
+  const displayDenom = useDisplayDenom(pluginId, currencyCode)
+  const exchangedFlipInputRef = React.useRef<ExchangedFlipInputRef>(null)
+  const feeCurrencyCode = getCurrencyCode(wallet, feeTokenId)
+  const feeDisplayDenom = useDisplayDenom(pluginId, feeCurrencyCode)
+
+  const [amounts, setAmounts] = useState<ExchangedFlipInputAmounts>({
+    nativeAmount: '',
+    exchangeAmount: '',
+    fiatAmount: '',
+    fieldChanged: forceField ?? 'fiat'
+  })
+
+  const walletName = wallet?.name ?? s.strings.string_no_name
+  const flipInputHeaderText = headerText ?? sprintf(s.strings.send_from_wallet, walletName)
+  const theme = useTheme()
+  const styles = getStyles(theme)
+
+  const handleAmountsChanged = useHandler((amounts: ExchangedFlipInputAmounts) => {
+    setAmounts(amounts)
+
+    if (onAmountsChanged) onAmountsChanged(amounts)
+  })
+
+  const handleFeesChange = useHandler(() => {
+    if (onFeesChange != null) {
+      handleCloseModal()
+      onFeesChange()
+    }
+  })
+
+  const handleCloseModal = useHandler(() => {
+    let { nativeAmount, exchangeAmount, fiatAmount } = amounts
+    nativeAmount = nativeAmount === '' ? '0' : nativeAmount
+    exchangeAmount = exchangeAmount === '' ? '0' : exchangeAmount
+    fiatAmount = fiatAmount === '' ? '0' : fiatAmount
+    bridge.resolve(Promise.resolve({ nativeAmount, exchangeAmount, fiatAmount }))
+  })
+
+  const handleSendMaxAmount = useHandler(() => {
+    if (onMaxSet != null) {
+      onMaxSet()
+      handleCloseModal()
+    }
+  })
+
+  const renderErrorMessage = useHandler(() => {
+    const opacity = errorMessage == null ? 0 : 1
+    return (
+      <EdgeText numberOfLines={1} style={[styles.exchangeRateErrorText, { opacity }]}>
+        {errorMessage == null ? ' ' : errorMessage.split('\n')[0]}
+      </EdgeText>
+    )
+  })
```

Comment:
Do not use `useHandler` for render helpers like this. They never get passed anywhere, and therefore we do not need to preserve their `===` identity changes.

Reply:
I use `useHandler` to prevent redeclaring the function on every component function call. My understanding is that this  makes the functions equivalent to class methods which are never redeclared. It's not as big of an optimization as functions that are passed as props, but still something

Reply:
The `useHandler` hook does not prevent that. The function still gets re-declared, but then `useHandler` wraps it in a "bouncer" function that retains it's `===` identity. So you still pay for the function to get re-declared, and then pay again for this bouncer we don't need.